ts/WoltLabSuite/WebComponent/woltlab-core-time.ts

```diff
@@ -0,0 +1,155 @@
+    Minutes: new Intl.RelativeTimeFormat(locale),
+    TodayOrYesterday: new Intl.RelativeTimeFormat(locale, { numeric: "auto" }),
+  };
+
+  const enum TimePeriod {
+    OneMinute = 60,
+    OneHour = 3_600,
+    OneDay = 86_400,
+    OneWeek = 86_400 * 7,
+  }
+
+  class WoltlabCoreTimeElement extends HTMLElement {
+    #date?: Date;
+    #timeElement?: HTMLTimeElement;
+
+    get date(): Date {
+      if (this.#date === undefined) {
+        const value = this.getAttribute("date");
+        if (!value) {
+          throw new Error("The 'date' attribute is missing.");
+        }
+
+        this.#date = new Date(value);
+      }
+
+      return this.#date;
+    }
+
+    set date(date: Date) {
+      this.setAttribute("date", date.toISOString());
+
+      this.refresh(true);
+    }
+
+    connectedCallback() {
+      this.refresh(true);
+    }
+
+    refresh(updateTitle: boolean) {
+      const date = this.date;
+      const difference = Math.trunc((Date.now() - date.getTime() - drift) / 1_000);
+
+      if (this.#timeElement === undefined) {
+        this.#timeElement = document.createElement("time");
+        this.#timeElement.dateTime = date.toISOString();
+
+        const shadow = this.attachShadow({ mode: "open" });
+        shadow.append(this.#timeElement);
+      }
+
+      if (updateTitle) {
+        this.#timeElement.title = DateFormatter.DateAndTime.format(date);
+      }
+
+      let value: string;
+      if (difference < TimePeriod.OneMinute) {
+        value = "TODO: a moment ago"; // Language.get("wcf.date.relative.now");
+      } else if (difference < TimePeriod.OneHour) {
+        const minutes = Math.trunc(difference / TimePeriod.OneMinute);
+        value = DateFormatter.Minutes.format(minutes * -1, "minute");
+      } else if (date.getTime() > todayDayStart) {
+        value = this.#formatTodayOrYesterday(date, TodayOrYesterday.Today);
+      } else if (date.getTime() > yesterdayDayStart) {
+        value = this.#formatTodayOrYesterday(date, TodayOrYesterday.Yesterday);
+      } else if (difference < TimePeriod.OneWeek) {
+        value = DateFormatter.DayOfWeekAndTime.format(date);
+      } else {
+        value = DateFormatter.Date.format(date);
+      }
+
+      value = value.charAt(0).toUpperCase() + value.slice(1);
+
+      this.#timeElement.textContent = value;
+    }
+
+    #formatTodayOrYesterday(date: Date, dayOffset: TodayOrYesterday): string {
+      let value = DateFormatter.TodayOrYesterday.format(dayOffset, "day");
+      const dateParts = DateFormatter.DayOfWeekAndTime.formatToParts(date);
+      if (dateParts[0].type === "weekday") {
+        const datePartsWithoutDayOfWeek: string[] = dateParts.slice(1).map((part) => part.value);
+        datePartsWithoutDayOfWeek.unshift(value);
+        value = datePartsWithoutDayOfWeek.join("");
+      }
+
+      return value;
+    }
+  }
+
+  window.customElements.define("woltlab-core-time", WoltlabCoreTimeElement);
+
+  const refreshAllTimeElements = function () {
```

Comment:
I did, but https://eslint.org/docs/latest/rules/no-inner-declarations appears to not play nicely with my use of `{ … }` instead of an IIFE. So I need to either use an IIFE or use a lambda function.

Reply:
Okay, fair enough.